File: src/backend/models/rfq.py
import enum
import uuid
from datetime import datetime
from decimal import Decimal

from sqlalchemy import DateTime, ForeignKey, Numeric, String, Text, func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from src.backend.db.base import Base


class RFQStatus(enum.Enum):
    DRAFT = "draft"
    SENT = "sent"
    RESPONDED = "responded"
    COMPARED = "compared"
    AWARDED = "awarded"
    CLOSED = "closed"
    CANCELLED = "cancelled"


class RFQ(Base):
    __tablename__ = "rfqs"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    project_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("projects.id"), nullable=False, index=True
    )
    vendor_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("vendors.id"), nullable=False, index=True
    )
    status: Mapped[str] = mapped_column(
        String(20), nullable=False, default=RFQStatus.DRAFT.value
    )
    rfq_number: Mapped[str] = mapped_column(
        String(50), nullable=False, unique=True, index=True
    )
    created_by: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=False
    )
    sent_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    responded_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    awarded_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    deleted_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )

    items: Mapped[list["RFQItem"]] = relationship(
        "RFQItem", back_populates="rfq", lazy="selectin"
    )


class RFQItem(Base):
    __tablename__ = "rfq_items"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    rfq_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("rfqs.id"), nullable=False, index=True
    )
    material_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("materials.id"), nullable=False
    )
    quantity: Mapped[Decimal] = mapped_column(Numeric(18, 2), nullable=False)
    vendor_rate: Mapped[Decimal | None] = mapped_column(Numeric(18, 2), nullable=True)
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)

    rfq: Mapped["RFQ"] = relationship("RFQ", back_populates="items")

File: src/frontend/src/types/rfq.ts
export type RFQStatus = "draft" | "sent" | "responded" | "compared" | "awarded" | "closed" | "cancelled";

export interface RFQ {
  id: string;
  project_id: string;
  vendor_id: string;
  vendor_name: string;
  status: RFQStatus;
  rfq_number: string;
  created_by: string;
  created_by_name: string;
  sent_at?: string;
  responded_at?: string;
  awarded_at?: string;
  notes?: string;
  created_at: string;
  items: RFQItem[];
}

export interface RFQItem {
  id: string;
  material_id: string;
  material_name: string;
  material_unit: string;
  quantity: number;
  vendor_rate?: number;
  notes?: string;
}

export interface RFQCompareItem {
  material_id: string;
  material_name: string;
  unit: string;
  quantity: number;
  vendor_rates: {
    rfq_id: string;
    vendor_id: string;
    vendor_name: string;
    rate: number;
  }[];
}

export interface RFQListResponse {
  items: RFQ[];
  total: number;
  page: number;
  page_size: number;
}

export interface RFQCreatePayload {
  project_id?: string;
  vendor_id: string;
  notes?: string;
  items: {
    material_id: string;
    quantity: number;
    notes?: string;
  }[];
}

export interface RFQRespondPayload {
  items: {
    item_id: string;
    vendor_rate: number;
    notes?: string;
  }[];
}

export interface RFQAwardPayload {
  notes?: string;
}
